### PR TITLE
Implement `StoreDriver::list` for `RedisStore`

### PR DESCRIPTION
**Pull request by barrbrain**
# Description

Advances #901 by enabling the locality-sensitive-hash store to be persisted and queried in Redis.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Adapted unit tests from the implementation for `MemoryStore`.

## Checklist

- [x] Updated documentation if needed
- [x] Tests added/amended
- [x] `bazel test //...`  passes locally
- [x] PR is contained in a single commit, using `git amend` see some [docs](https://www.atlassian.com/git/tutorials/rewriting-history)

- - -
This change is [Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697)

### CODE REVIEW

**Comment by MarcusSorealheis** [comment]
Thanks for starting this effort. I'll be following along and watching closely. One thing that gives me heartburn, is the rumor in the company that there are some non-standard APIs in use in our Redis Store.  I don't have the details. 

We don't see an issue in our cloud, but I will keep an eye on this effort and look to test the functionality as soon as it is testable.

**Comment by MarcusSorealheis** [comment]
And honestly, looks good so far!

**Comment by barrbrain** [comment]
For my use case, `list()` provides a convenient hook for scanning keys with a given prefix. `MemoryStore::list()` was good enough to prototype with but lacked persistence. Depending on the store implementation, this query pattern may be expensive.

**Review by aaronmondal — CHANGES_REQUESTED** [review]
The bound-based inputs seem quite elegant!

+@jhpratt

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-:-OOdYvrVAUXlPpAIBU8R:b-d39t0g)* status: 0 of 2 LGTMs obtained, and all files reviewed, and 2 discussions need to be resolved (waiting on @jhpratt)

___
*[`nativelink-store/src/redis_store.rs` line 397 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-OOdYXa78ump6_ZPc86Y:-OOdYXa78ump6_ZPc86Z:b-tq96qg) ([raw file](https://github.com/TraceMachina/nativelink/blob/154ead4610b2e0b116eb183cebbc77f9173213ee/nativelink-store/src/redis_store.rs#L397)):*
> ```XML
>         };
>         let client = self.client_pool.next();
>         let mut scan_stream = client.scan(pattern, Some(10000), None);
> ```

The `1000` value should probably either be a named const, or potentially even a value that is configurable via the config.
___
*[`nativelink-store/tests/redis_store_test.rs` line 936 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-OOd_9l4Efw6lJUaSzvV:-OOd_9l4Efw6lJUaSzvW:bv3ddjk) ([raw file](https://github.com/TraceMachina/nativelink/blob/154ead4610b2e0b116eb183cebbc77f9173213ee/nativelink-store/tests/redis_store_test.rs#L936)):*
> ```XML
>         let keys = get_list(&store, KEY2..KEY3).await;
>         assert_eq!(keys, vec![KEY2]);
>     }
> ```

What happens in the case of invalid inputs?

**Review by barrbrain — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697)* status: 0 of 2 LGTMs obtained, and 0 of 4 files reviewed, and pending CI: Analyze (javascript-typescript), Analyze (python), Publish image, Web Platform Deployment / macos-15, Web Platform Deployment / ubuntu-24.04, buildstream, pre-commit-checks, ubuntu-24.04 / stable, and 2 discussions need to be resolved (waiting on @jhpratt)

___
*[`nativelink-store/src/redis_store.rs` line 397 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-OOdYXa78ump6_ZPc86Y:-OOdmnHP7dL9ihjvPHH2:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/154ead4610b2e0b116eb183cebbc77f9173213ee/nativelink-store/src/redis_store.rs#L397)):*
<details><summary><i>Previously, aaronmondal (Aaron Siddhartha Mondal) wrote…</i></summary><blockquote>

The `1000` value should probably either be a named const, or potentially even a value that is configurable via the config.
</blockquote></details>

Done.
___
*[`nativelink-store/tests/redis_store_test.rs` line 936 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-OOd_9l4Efw6lJUaSzvV:-OOdmxWL80ru7qdn_YB8:b5sedk7) ([raw file](https://github.com/TraceMachina/nativelink/blob/154ead4610b2e0b116eb183cebbc77f9173213ee/nativelink-store/tests/redis_store_test.rs#L936)):*
<details><summary><i>Previously, aaronmondal (Aaron Siddhartha Mondal) wrote…</i></summary><blockquote>

What happens in the case of invalid inputs?
</blockquote></details>

There are 2 modes of invalid inputs:
* Invalid range bounds - work is done but results always empty
* Invalid keys from Redis - silently dropped if there is a conversion error

**Review by barrbrain — COMMENTED** [review]
Thank you for the thoughtful review. This is looking more polished than my original draft.

> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-:-OOiQNow-VTVlUL2ihCC:b-7qacj7)* status: 0 of 2 LGTMs obtained, and 2 of 4 files reviewed, and pending CI: Analyze (javascript-typescript), Analyze (python), Bazel Dev / macos-15, Bazel Dev / ubuntu-24.04, Cargo Dev / macos-15, Cargo Dev / ubuntu-24.04, Coverage, Installation / macos-13, Installation / macos-14, Installation / macos-15, Installation / ubuntu-22.04, Installation / ubuntu-24.04, Local / lre-cc / ubuntu-24.04, Local / lre-rs / macos-15, Local / lre-rs / ubuntu-24.04, NativeLink.com Cloud / Remote Cache / macos-15, NativeLink.com Cloud / Remote Cache / ubuntu-24.04, Publish image, Publish nativelink-worker-init, Publish nativelink-worker-lre-cc, Remote / lre-cc / large-ubuntu-22.04, Remote / lre-rs / large-ubuntu-22.04, Web Platform Deployment / macos-15, Web Platform Deployment / ubuntu-24.04, asan / ubuntu-24.04, buildstream, integration-tests (24.04), macos-15, pre-commit-checks, ubuntu-24.04, ubuntu-24.04 / stable, vale, windows-2022 / stable, and 3 discussions need to be resolved (waiting on @jhpratt)

___
*[`nativelink-store/src/redis_store.rs` line 280 at r3](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-OOdpcyz8zlzuZmMtSrj:-OOiPoth2D6HI4zeSRJg:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/50f090f80ff4ebc7a4af3671f0135bcf94422bf7/nativelink-store/src/redis_store.rs#L280)):*
<details><summary><i>Previously, jhpratt (Jacob Pratt) wrote…</i></summary><blockquote>

Let's use `#[expect]` so that it warns us if this is no longer needed.
</blockquote></details>

Done.

**Review by aaronmondal — COMMENTED** [review]
> Reviewed 2 of 4 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697)* status: 0 of 2 LGTMs obtained, and all files reviewed, and pending CI: Bazel Dev / macos-15, Bazel Dev / ubuntu-24.04, Cargo Dev / macos-15, Cargo Dev / ubuntu-24.04, Coverage, Installation / macos-13, Installation / macos-14, Installation / macos-15, Installation / ubuntu-22.04, Installation / ubuntu-24.04, Local / lre-rs / macos-15, NativeLink.com Cloud / Remote Cache / macos-15, NativeLink.com Cloud / Remote Cache / ubuntu-24.04, Publish image, Publish nativelink-worker-init, Remote / lre-cc / large-ubuntu-22.04, Remote / lre-rs / large-ubuntu-22.04, Web Platform Deployment / macos-15, buildstream, macos-15, windows-2022 / stable, and 2 discussions need to be resolved (waiting on @jhpratt)

___
*[`nativelink-store/tests/redis_store_test.rs` line 936 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-OOd_9l4Efw6lJUaSzvV:-OOiUeyu1I-O22YUZf_W:bjl763t) ([raw file](https://github.com/TraceMachina/nativelink/blob/154ead4610b2e0b116eb183cebbc77f9173213ee/nativelink-store/tests/redis_store_test.rs#L936)):*
<details><summary><i>Previously, barrbrain (David Michael Barr) wrote…</i></summary><blockquote>

There are 2 modes of invalid inputs:
* Invalid range bounds - work is done but results always empty
* Invalid keys from Redis - silently dropped if there is a conversion error
</blockquote></details>

Could we either add a TODO to test these cases or implement tests for this? IIUC the first case would be a case that returns an empty result. Not sure about the second case. Silently dropping any sort of failure seems a bit risky to me, but if that's hard to test we could maybe at least leave a docstring or TODO somewhere to make sure this isn't ovelooked in the future.
___
*[`nativelink-scheduler/tests/redis_store_awaited_action_db_test.rs` line 54 at r4](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-OOiTMjl9orryGblw08W:-OOiTMjl9orryGblw08X:bau6ldw) ([raw file](https://github.com/TraceMachina/nativelink/blob/977e1e852e004903794dc5438786d7f35c3ab4e0/nativelink-scheduler/tests/redis_store_awaited_action_db_test.rs#L54)):*
> ```XML
> const VERSION_SCRIPT_HASH: &str = "fdf1152fd21705c8763752809b86b55c5d4511ce";
> const MAX_CHUNK_UPLOADS_PER_UPDATE: usize = 10;
> const SCAN_COUNT: u32 = 10_000;
> ```

fyi: I initially thought that we could omit this, but that doesn't seem to be the case because `new_from_builder_and_parts` doesn't respect the defaults. Only `new` does.

@jhpratt shouldn't the defaults be set for some sort of `RedisStore::default` implementation instead of `RedisStore::new`? But I guess that would complicate callsites. Is there some common way to handle this kind of thing? Initially, the current behavior here (and likely in most other stores) seems unintuitive to me (but I might be wrong). Also, aren't we essentially misusing the builder pattern for the construction of these stores (and like basically every other store :sweat_smile:)?

(Not related to this PR)

**Review by barrbrain — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697)* status: 0 of 2 LGTMs obtained, and 2 of 4 files reviewed, and pending CI: Analyze (javascript-typescript), Analyze (python), Bazel Dev / macos-15, Bazel Dev / ubuntu-24.04, Coverage, Installation / macos-13, Installation / macos-14, Installation / macos-15, Installation / ubuntu-22.04, Installation / ubuntu-24.04, Local / lre-cc / ubuntu-24.04, Local / lre-rs / macos-15, Local / lre-rs / ubuntu-24.04, NativeLink.com Cloud / Remote Cache / macos-15, NativeLink.com Cloud / Remote Cache / ubuntu-24.04, Publish image, Publish nativelink-worker-init, Publish nativelink-worker-lre-cc, Remote / lre-cc / large-ubuntu-22.04, Remote / lre-rs / large-ubuntu-22.04, Web Platform Deployment / macos-15, Web Platform Deployment / ubuntu-24.04, asan / ubuntu-24.04, buildstream, integration-tests (24.04), macos-15, pre-commit-checks, ubuntu-24.04, ubuntu-24.04 / stable, vale, windows-2022 / stable, and 2 discussions need to be resolved (waiting on @jhpratt)

___
*[`nativelink-store/tests/redis_store_test.rs` line 936 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-OOd_9l4Efw6lJUaSzvV:-OOicSBG9kam-OapNZ7y:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/154ead4610b2e0b116eb183cebbc77f9173213ee/nativelink-store/tests/redis_store_test.rs#L936)):*
<details><summary><i>Previously, aaronmondal (Aaron Siddhartha Mondal) wrote…</i></summary><blockquote>

Could we either add a TODO to test these cases or implement tests for this? IIUC the first case would be a case that returns an empty result. Not sure about the second case. Silently dropping any sort of failure seems a bit risky to me, but if that's hard to test we could maybe at least leave a docstring or TODO somewhere to make sure this isn't ovelooked in the future.
</blockquote></details>

Done.

**Review by aaronmondal — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-:-OOic_6V8D9863GzEl7z:bnfp4nl)* status: 1 of 2 LGTMs obtained, and all files reviewed, and pending CI: Bazel Dev / macos-15, Bazel Dev / ubuntu-24.04, Cargo Dev / macos-15, Cargo Dev / ubuntu-24.04, Coverage, Installation / macos-13, Installation / macos-14, Installation / macos-15, Installation / ubuntu-22.04, Installation / ubuntu-24.04, NativeLink.com Cloud / Remote Cache / macos-15, NativeLink.com Cloud / Remote Cache / ubuntu-24.04, Publish image, Publish nativelink-worker-init, Remote / lre-cc / large-ubuntu-22.04, Remote / lre-rs / large-ubuntu-22.04, Web Platform Deployment / macos-15, Web Platform Deployment / ubuntu-24.04, buildstream, windows-2022 / stable, and 1 discussions need to be resolved (waiting on @jhpratt)

**Review by jhpratt — APPROVED** [review]
:lgtm:

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-:-OOiou213PpOTxtHyyOs:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 2 LGTMs obtained, and all files reviewed

___
*[`nativelink-scheduler/tests/redis_store_awaited_action_db_test.rs` line 54 at r4](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-OOiTMjl9orryGblw08W:-OOinDXACbDq6tskUde0:b-2bcoiu) ([raw file](https://github.com/TraceMachina/nativelink/blob/977e1e852e004903794dc5438786d7f35c3ab4e0/nativelink-scheduler/tests/redis_store_awaited_action_db_test.rs#L54)):*
<details><summary><i>Previously, aaronmondal (Aaron Siddhartha Mondal) wrote…</i></summary><blockquote>

fyi: I initially thought that we could omit this, but that doesn't seem to be the case because `new_from_builder_and_parts` doesn't respect the defaults. Only `new` does.

@jhpratt shouldn't the defaults be set for some sort of `RedisStore::default` implementation instead of `RedisStore::new`? But I guess that would complicate callsites. Is there some common way to handle this kind of thing? Initially, the current behavior here (and likely in most other stores) seems unintuitive to me (but I might be wrong). Also, aren't we essentially misusing the builder pattern for the construction of these stores (and like basically every other store :sweat_smile:)?

(Not related to this PR)
</blockquote></details>

In a way, yes. The builder pattern requires a fair amount of boilerplate unless macros are used (and they're not in this codebase). There's an upcoming language feature that will largely replace the builder pattern _and_ work with default values seamlessly, but it's not stabilized yet.

I have noticed that the default values are not actually respected in many parts of the codebase, but I think for this PR we should at least be consistent with that.

**Comment by barrbrain** [comment]
The tests should probably include a non-matching case. There appears to be a regression in `fred` for handling the empty response, since the version I used for private testing.

File: nativelink-store/src/redis_store.rs [hunk 1]
```diff
@@ -365,6 +383,58 @@ impl StoreDriver for RedisStore {
...
+                    format!("{}{}*", self.key_prefix, &start[..length])
+                }
+                Bound::Unbounded => format!("{}*", self.key_prefix),
+            },
+            Bound::Unbounded => format!("{}*", self.key_prefix),
+        };
+        let client = self.client_pool.next();
+        let mut scan_stream = client.scan(pattern, Some(self.scan_count), None);
+        let mut iterations = 0;
+        'outer: while let Some(mut page) = scan_stream.try_next().await? {
```
Comment (barrbrain) [inline nativelink-store/src/redis_store.rs:415]: The unexpected protocol error from `fred` may be suppressed but then there is no path for errors to propagate.
```suggestion
        'outer: while let Some(mut page) = scan_stream.try_next().await.unwrap_or(None) {
```

Comment (barrbrain) [reply nativelink-store/src/redis_store.rs:415]: I confirmed that backporting this implementation to `v0.5.3` works as expected. That release's `Cargo.lock` is using `fred 9.1.2`, which precedes the changes to scan in `fred 10.0.0`. I will try to isolate the issue and make a bug report.

Comment (MarcusSorealheis) [reply nativelink-store/src/redis_store.rs:415]: Oh, great. I'm sure the `fred` devs will appreciate it. As you know, `0.5.x` in NativeLink is not advised. Maybe we can downgrade the `fred` version, depending on why we upgraded it. 

Thank you.

Comment (barrbrain) [reply nativelink-store/src/redis_store.rs:415]: Okay, the actual reason is embarrassing. I missed adding the logic for default `scan_count`.

Comment (MarcusSorealheis) [reply nativelink-store/src/redis_store.rs:415]: It happens.

**Review by aaronmondal — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697#-:-OP6GJ435yJxRZaskkiy:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 2 LGTMs obtained, and all files reviewed

**Review by aaronmondal — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/1697)* status: <span>:shipit:</span> complete! 2 of 2 LGTMs obtained, and all files reviewed
